internal/proxy/deltaproxycache.go

```diff
@@ -77,80 +105,149 @@ func DeltaProxyCacheRequest(r *Request, w http.ResponseWriter, client Client, ca
-		Respond(w, cacheData.StatusCode, cacheData.Headers, cacheData.Body)
-		return
+		if cacheStatus == crPartialHit {
+			cacheStatus = crHit
+		}
+	} else if len(missRanges) == 1 && missRanges[0].Start == trq.Extent.Start && missRanges[0].End == trq.Extent.End {
+		cacheStatus = crRangeMiss
+	}
+
+	// Increment Counters
+	metrics.ProxyRequestStatus.WithLabelValues(r.OriginName, r.OriginType, r.HTTPMethod, cacheStatus, strconv.Itoa(doc.StatusCode), r.URL.Path).Inc()
+	metrics.ProxyRequestDuration.WithLabelValues(r.OriginName, r.OriginType, r.HTTPMethod, cacheStatus, strconv.Itoa(doc.StatusCode), r.URL.Path).Observe(elapsed.Seconds())
+
+	if len(missRanges) == 0 {
+		if cfg.FastForwardDisable {
+			Respond(w, doc.StatusCode, doc.Headers, doc.Body)
+			return
+		}
 	}
 
 	// maintain a list of timeseries to merge into the main timeseries
 	mts := make([]timeseries.Timeseries, 0, len(missRanges))
 	wg := sync.WaitGroup{}
 	appendLock := sync.Mutex{}
 	var rh http.Header
+
+	// iterate each time range that the client needs and fetch from the upstream origin
 	for i := range missRanges {
 		wg.Add(1)
 		req := r.Copy() // copy the request headers so we avoid collisions when adjusting them
+
 		// This fetches the gaps from the origin and adds their datasets to the merge list
 		go func(e *timeseries.Extent, r *Request) {
 			defer wg.Done()
 			client.SetExtent(req, e)
-			body, resp, elapsed := Fetch(req)
+			body, resp, _ := Fetch(req)
 			if resp.StatusCode == http.StatusOK && len(body) > 0 {
 				nts, err := client.UnmarshalTimeseries(body)
 				if err != nil {
 					log.Error("proxy object unmarshaling failed", log.Pairs{"body": string(body)})
 					return
 				}
-
-				cacheStatus := "phit"
-				if e.Start == trq.Extent.Start && e.End == trq.Extent.End {
-					cacheStatus = "rmiss"
-				}
-
 				nts.SetExtents([]timeseries.Extent{*e})
-				metrics.ProxyRequestStatus.WithLabelValues(req.OriginName, req.OriginType, req.HTTPMethod, cacheStatus, strconv.Itoa(resp.StatusCode), req.URL.Path).Inc()
-				metrics.ProxyRequestDuration.WithLabelValues(req.OriginName, req.OriginType, req.HTTPMethod, cacheStatus, strconv.Itoa(resp.StatusCode), req.URL.Path).Observe(elapsed.Seconds())
 				appendLock.Lock()
 				defer appendLock.Unlock()
 
 				mts = append(mts, nts)
 				if rh == nil {
+					// Use the response headers from the first successful delta request to complete as our downstream response headers
 					rh = resp.Header
 				}
 			}
 		}(&missRanges[i], req)
 	}
 
-	// TODO: Fast Forward Here, Another wg.Add(1), go func, defer wg.Done(), fetch, Matrix from Vector, lock, defer unlock, append to mts
+	var ffts timeseries.Timeseries
+	var ffex timeseries.Extent
+	// Only fast forward if configured and the user request is for the absolute latest datapoint
+	if (!cfg.FastForwardDisable) && (trq.Extent.End == normalizedNow.Extent.End) {
+		ffURL, err := client.FastForwardURL(r)
+		if ffURL.Scheme != "" && err == nil {
+			wg.Add(1)
+			go func() {
```

Comment:
Is there any chance this will block forever? Does having a timeout on this make sense?

Reply:
Should be set now on that, the client timeout is a feature in 0.1.x that i neglected to port over, so thank you :)  It is now passed all the way through to proxy.Fetch() on a per-origin basis.